Src/Common/InstrumentationKeyHashLookupHelper.cs

```diff
@@ -0,0 +1,73 @@
+﻿namespace Microsoft.ApplicationInsights.Common
+{
+    using System;
+    using System.Collections.Concurrent;
+    using System.Globalization;
+    using System.Security.Cryptography;
+    using System.Text;
+
+    /// <summary>
+    /// A store for instrumentation key hashes. This is order so as to optimize the computation of hashes.
+    /// </summary>
+    internal static class InstrumentationKeyHashLookupHelper
+    {
+        /// <summary>
+        /// Max number of component hashes to cache.
+        /// </summary>
+        private const int MAXSIZE = 100;
+
+        private static ConcurrentDictionary<string, string> knownIKeyHashes = new ConcurrentDictionary<string, string>();
+
+        /// <summary>
+        /// Retrieves the hash for a given instrumentation key.
+        /// </summary>
+        /// <param name="instrumentationKey">Instrumentation key string.</param>
+        /// <returns>SHA256 hash for the instrumentation key.</returns>
+        public static string GetInstrumentationKeyHash(string instrumentationKey)
+        {
+            if (string.IsNullOrEmpty(instrumentationKey))
+            {
+                throw new ArgumentNullException("instrumentationKey");
+            }
+
+            string hash;
+            var found = knownIKeyHashes.TryGetValue(instrumentationKey, out hash);
+
+            if (!found)
+            {
+                var instrumentationKeyLowerCase = instrumentationKey.ToLowerInvariant();
+                
+                // Instrumentation key is not case sensitive, let's now try for the lowercase - may be that's found.
+                found = knownIKeyHashes.TryGetValue(instrumentationKeyLowerCase, out hash);
```

Comment:
If you looking it up in dictionary - add it to the dictionary as well. Otherwise don't do another lookup